### PR TITLE
Replace Cassette for an overlayed `MethodTable`

### PR DESCRIPTION
**Pull request by mofeing**
CC @wsmoses @vchuravy

~~The only remaining thing is an alternative implementation of `Base.invoke_within` to effectively overlaying the methods with our `MethodTable`.~~

Also, we might need to add some definitions based on Julia version due to changes in the `AbstractInterpreter` API.

- [x] Fix `OpaqueClosure` call on Julia 1.9
- [x] Fix `TypeError: in new, expected DataType, got Type{Symbol}` on Julia 1.11
- [ ] Support kwargs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,7 +8,7 @@
...
+Reactant takes Julia function and compile it into MLIR and run fancy optimizations on top of it, including using EnzymeMLIR for automatic differentiation, and create relevant executables for CPU/GPU/TPU via XLA. Compiled functions will assume the same control flow pattern as was original taken by objects used at compile time, and control flow (e.g. if, for) as well as any type instabilities will be removed. The benefits of this approach is immediately making all such code available for advanced optimization with little developer effort. This system and corresponding semantics is subject to change to a (potentially partial) source rewriter in the future.
```
Comment (wsmoses) [inline README.md:11]: We should still keep the sentence "It presently operates as a tracing system" even if no casette term

File: src/Interpreter.jl [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+            continue
+        end
+        if path[1] == :result
+            return true
+        end
+    end
+    return false
+end
+
+@hlo_override function Enzyme.autodiff(
```
Comment (wsmoses) [inline src/Interpreter.jl:211]: minor, but lets call the macro reactant_override

Comment (mofeing) [reply src/Interpreter.jl:211]: Okay 👍

**Comment by mofeing** [comment]
@wsmoses This seems like a problem of Enzyme.jl on Julia 1.11 https://github.com/EnzymeAD/Reactant.jl/actions/runs/9944022931/job/27469120868?pr=40#step:9:561

**Comment by wsmoses** [comment]
Oh yea, but that's an easy fix here, we should just not run the non-reactant autodiff code on Julia <= 1.10

**Comment by mofeing** [comment]
In Julia <= 1.10? Wouldn't it be in Julia >=1.11?

**Comment by wsmoses** [comment]
Oh yeah sorry that’s what I meant

**Comment by mofeing** [comment]
Okay, but where does that happen? I mean, we are overriding `Enzyme.autodiff` so the original method shouldn't get called in `Reactant` right?

File: test/basic.jl [hunk 3]
```diff
@@ -146,9 +146,10 @@ end
 
 @testset "Reactant.@code_hlo" begin
     W = Reactant.ConcreteRArray(randn(Float32, 10, 20))
-    x = Reactant.ConcreteRArray(randn(Float32, 20, 5))
-    res = Reactant.@code_hlo W * x
+    # x = Reactant.ConcreteRArray(randn(Float32, 20, 5))
+    # res = Reactant.@code_hlo W * x
```
Comment (wsmoses) [inline test/basic.jl:150]: unrelated change, no?

**Review by wsmoses — APPROVED** [review]
11/10

File: test/basic.jl (hunk 3 above)
Comment (mofeing) [reply test/basic.jl:150]: Apparently, `@code_hlo W * x` crashes in Julia 1.9 with our interpreter 🤷

**Comment by mofeing** [comment]
I think I found the problem for the crash on `@code_hlo W * x`. It seems like the return type for `Base.:*` with `TracedRArray`s is type-unstable, and the interpreter doesn't like that on Julia 1.9 (still type-unstable on Julia 1.10 but don't know why it accepts it 🤷).

# Comparison
## Julia 1.9

```julia
julia> code_warntype(*, (Reactant.TracedRArray{Float32, (10,20), 2}, Reactant.TracedRArray{Float32, (20,5), 2}))
MethodInstance for *(::Reactant.TracedRArray{Float32, (10, 20), 2}, ::Reactant.TracedRArray{Float32, (20, 5), 2})
  from *(lhs::Reactant.TracedRArray{ElType, Shape, 2}, rhs::Reactant.TracedRArray{ElType, Shape2, 2}) where {ElType, Shape, Shape2} @ Reactant /media/M2/mofeing/repos/Reactant.jl/src/overloads.jl:143
Static Parameters
  ElType = Float32
  Shape = (10, 20)
  Shape2 = (20, 5)
Arguments
  #self#::Core.Const(*)
  lhs::Reactant.TracedRArray{Float32, (10, 20), 2}
  rhs::Reactant.TracedRArray{Float32, (20, 5), 2}
Locals
  res::Reactant.MLIR.IR.Value
  precar::Reactant.MLIR.IR.Attribute
  prec::Reactant.MLIR.IR.Attribute
  dot_dimension_numbers::Reactant.MLIR.API.MlirAttribute
  resty::Reactant.MLIR.IR.Type
  rhsty::Reactant.MLIR.IR.Type
  lhsty::Reactant.MLIR.IR.Type
Body::Reactant.TracedRArray{Float32, _A, 2} where _A
1 ─ %1  = Reactant.MLIR.IR::Core.Const(Reactant.MLIR.IR)
│   %2  = Base.getproperty(%1, :type)::Core.Const(Reactant.MLIR.IR.type)
...
│   %67 = (%65)(%66, res)::Reactant.TracedRArray{Float32, _A, 2} where _A
└──       return %67
```

## Julia 1.9 with new implementation

```julia
julia> code_warntype(*, (Reactant.TracedRArray{Float32, (10,20), 2}, Reactant.TracedRArray{Float32, (20,5), 2}))
MethodInstance for *(::Reactant.TracedRArray{Float32, (10, 20), 2}, ::Reactant.TracedRArray{Float32, (20, 5), 2})
  from *(lhs::Reactant.TracedRArray{ElType, Shape, 2}, rhs::Reactant.TracedRArray{ElType, Shape2, 2}) where {ElType, Shape, Shape2} @ Reactant /media/M2/mofeing/repos/Reactant.jl/src/overloads.jl:143
Static Parameters
  ElType = Float32
  Shape = (10, 20)
  Shape2 = (20, 5)
Arguments
  #self#::Core.Const(*)
  lhs::Reactant.TracedRArray{Float32, (10, 20), 2}
  rhs::Reactant.TracedRArray{Float32, (20, 5), 2}
Locals
  res::Reactant.MLIR.IR.Value
  precar::Reactant.MLIR.IR.Attribute
  prec::Reactant.MLIR.IR.Attribute
  dot_dimension_numbers::Reactant.MLIR.API.MlirAttribute
  resty::Reactant.MLIR.IR.Type
  rhsty::Reactant.MLIR.IR.Type
  lhsty::Reactant.MLIR.IR.Type
Body::Reactant.TracedRArray{Float32, (10, 5), 2}
1 ─ %1  = Reactant.MLIR.IR::Core.Const(Reactant.MLIR.IR)
│   %2  = Base.getproperty(%1, :type)::Core.Const(Reactant.MLIR.IR.type)
...
│   %63 = (%61)(%62, res)::Reactant.TracedRArray{Float32, (10, 5), 2}
└──       return %63
```

**Comment by mofeing** [comment]
The kwargs support can be postponed since `compile` doesn't support yet kwargs.
